src/main/java/pp/springdata/todo/TodoController.java

```diff
@@ -0,0 +1,89 @@
+package pp.springdata.todo;
+
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Controller
+public class TodoController {
+
+    private final TodoRepository todoRepository;
+
+    public TodoController(TodoRepository todoRepository) {
+        this.todoRepository = todoRepository;
+    }
+
+    @GetMapping("/")
+    public String home(Model model, @RequestParam(required = false) Category category,
+                       @RequestParam(required = false, defaultValue = "false") Boolean archive) {
+        List<Todo> todoList;
+        if (category != null) {
+            todoList = todoRepository.findByCategoryOrderByDeadline(category);
+        } else {
+            todoList = todoRepository.findAll(Sort.by("deadline"));
+        }
+
+        model.addAttribute("todoList", todoList.stream()
+                .filter(todo -> todo.getDone().equals(archive))
+                .collect(Collectors.toList()));
+        model.addAttribute("todoAdd", new Todo());
+        return "home";
+    }
+
+    @GetMapping("/todo/{id}")
+    public String showToDo(@PathVariable Long id, Model model) {
+        Optional<Todo> todoOptional = todoRepository.findById(id);
+
+        if (todoOptional.isPresent()) {
+            Todo todo = todoOptional.get();
+            model.addAttribute("todo", todo);
+            return "todo";
+        } else {
+            return "redirect:/";
+        }
+    }
+
+    @GetMapping("/todo/{id}/edit")
+    public String todoEditForm(@PathVariable Long id, Model model) {
+        Optional<Todo> todoOptional = todoRepository.findById(id);
+        if (todoOptional.isPresent()) {
+            Todo todo = todoOptional.get();
+            model.addAttribute("todoToEdit", todo);
+            return "todoEdit";
+        } else {
+            return "redirect:/";
+        }
+    }
+
+    @PostMapping("/todo/{id}/edit")
+    public String editTask(@PathVariable Long id, Todo todo) {
+        Todo todo1 = todoRepository.findById(id).orElseThrow();
```

Comment:
Jak już coś rzucasz to lepiej rzucić konkretny exception z jakąś wiadomością w środku - łatwiej potem dochodzić co się stało. orElseThrow przyjmuje jako jeden z paramterów Supplier gdzie możesz dać własny excepiton.